### ISSUE TITLE
Sharing a complete Realm using PermissionOffer and PermissionOfferResponse results in an assertion error

### ISSUE THREAD

**Issue by Taco55** [realm/realm-mobile-platform#136]
When I try to share a complete Realm between two users using `PermissionOffer` and `PermissionOfferResponse` I encounter several problems.

Although I succeed to share a `PermissionOffer` token with a user, and this receiving user is able to obtain the a realmURL using `PermissionOfferResponse`, it appears that the shared Realm is empty when opened by this receiving user. Furthermore, the second time a token is generated by the user's management Realm, opening the shared Realm results in an assertion error:

```
.../PermissionChangeDemo/Pods/Realm/Realm/RLMClassInfo.mm:97: [realm-core-2.3.0] Assertion failed: rlmSchema.objectSchema.count == schema.size()
0   Realm                               0x0000000107e018b8 _ZN5realm4utilL18terminate_internalERNSt3__118basic_stringstreamIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 40
1   Realm                               0x0000000107e01bba _ZN5realm4util9terminateEPKcS2_lOSt16initializer_listINS0_9PrintableEE + 506
2   Realm                               0x0000000107b4d8b0 _ZN13RLMSchemaInfoC2EP8RLMRealmP9RLMSchemaRKN5realm6SchemaE + 1008
3   Realm                               0x0000000107b4e8f3 _ZN13RLMSchemaInfoC1EP8RLMRealmP9RLMSchemaRKN5realm6SchemaE + 83
4   Realm                               0x0000000107c90449 _ZL25RLMRealmSetSchemaAndAlignP8RLMRealmP9RLMSchema + 185
5   Realm                               0x0000000107c92873 +[RLMRealm realmWithConfiguration:error:] + 5779
6   RealmSwift                          0x0000000108456402 _TTOFCSo8RLMRealmCfzT13configurationCSo21RLMRealmConfiguration_S_ + 146
7   RealmSwift                          0x0000000108456777 _TFC10RealmSwift5RealmcfzT13configurationVS0_13Configuration_S0_ + 743
8   RealmSwift                          0x0000000108456bce _TFC10RealmSwift5RealmCfzT13configurationVS0_13Configuration_S0_ + 558
9   PermissionChangeDemo                0x00000001079752ce 
```

## Expected Results
During the first run I would expect to get access to all the data in the shared Realm. During the second run after processing `PermissionOffer` and `PermissionOfferResponses` for the second time, I would not expect an assertion error after opening the shared Realm.

Unfortunately, it is not clear to me whether it is allowed to store multiple `PermissionOfferResponses` (for the same Realm that is shared) to the user's management Realm. For example, should it be determined whether a `PermissionOfferResponse` has already been processed? Or would a user be able to process any number of `PermissionOfferResponses`  (e.g. also for Realm's that have already been "shared")? In any case, a more helpful error message would be desirable instead of indicating an assertion error only.

## Code Sample
To illustrate the problem, I created a sample application. For the first run, the receiving user generates a realmURL, but when used to open the "shared Realm" it appears to be an empty Realm. For the second run, an assertion error occurs when the realmURL is used to open the "shared Realm".

See [this repository](https://github.com/Taco55/RealmPermissionChangeDemo) for a complete example.

## Version of Realm and Tooling

Realm Object Server Version: 1.0.3 (developer)
Realm Database Version: 2.4.3
Server OS & Version: Mac
Client OS & Version: 10.2

**Comment by istx25** [comment]
Hi @Taco55. Thanks for reaching out. Someone will review what you've provided and follow-up with more information or questions soon. Cheers!

**Comment by Taco55** [comment]
After continuing for a considerable time I still do not get the sharing working. Might this be a bug though? I would really appreciate any help on this.

Furthermore, I also encountered that `SyncPermissionOffer` sometimes does not generate a token after it is added to the user's management Realm. Despite removing derived data solves this issue, it feels strange that the derived data would influence a server's response. Is this something known? Unfortunately, I am unable to reproduce this issue.

I updated the [example](https://github.com/Taco55/RealmPermissionChangeDemo) somewhat to better illustrate the problems with sharing during running. Hopefully, this clear enough. Many thanks in advance.

**Comment by istx25** [comment]
My apologies for the delay in someone getting back to you. This issue got lost in our backlog but now that we have some eyes on the issue, someone will respond soon. Sorry!

**Comment by Taco55** [comment]
Although I understand that it may cost some time to review my problem, I would like to add that similar issues seems to occur with the _jp-permission-offer-response_ branch of RealmTasks.
 
For example, when a shared Realm of a list is opened with the `fullServerPath`, the Realm appears to be empty (e.g. list name is empty and there are no tasks). This is despite the fact that I could successfully store the fullServerPath in `TaskListReference`. 
Furthermore, it appears that the realmtasks link can be opened only once or it will crash. 

Can I conclude that it is allowed to add only one `SyncPermissionOfferResponse` object to a user's management Realm per each shared Realm? And is it possible to provide a global indication whether any changes/fixes to `SyncPermissionOffer` and `Response` could be expected in the next release? And is there also any time indication available?

In my case, it would be really great to get this functionality working as intended.

**Comment by austinzheng** [comment]
I apologize for the delay in getting back to you. I've run your reproduction project successfully, and am working on cutting your project to as minimal as possible in order to isolate what is going on.

The `jp-permission-offer-response` branch of RealmTasks is a work in progress branch and shouldn't be taken as an indicator as to what is intended and what isn't. As for your other questions someone who is more familiar with the permissions subsystem will have to help you out. (cc. @mrackwitz).

**Comment by austinzheng** [comment]
An [object store](https://github.com/realm/realm-object-store) bug was discovered that was causing the assert to be called. A fix for it should be committed today or tomorrow. Again, thank you so much for offering us a reproduction case! We will post updates to this thread as necessary.

> Can I conclude that it is allowed to add only one SyncPermissionOfferResponse object to a user's management Realm per shared Realm?

I'll ask our engineers who specialize in sync and get back to you with an answer.

**Comment by Taco55** [comment]
Thanks for your response and help on this. Also glad to hear that you were able to find the cause of this issue. Once the updates are committed to the repository I will test it in my application.

Keep up the good work!

**Comment by Taco55** [comment]
Since I still encountered problems with the permission sharing, it took a while before I worked on this again. At this moment I would like to report a small related issue and ask a question.  

First of all, the main issue related to the assertion error is solved and I succeeded to share a user's Realm! However, some problems occur after resetting the MacOS version of Realm object server. After a reset, no tokens are generated anymore after adding a `SyncPermissionOffer` to the user's management Realm. More specifically, after adding a `addNotificationBlock` to a `SyncPermissionOffer` collection, only an `.initial` change is observed while an .`update` (with an expected token) does not occur. 
A solution seemed to re-install the server everytime I want an empty database. Since I prefer to refresh the database this is not ideal. Should I add this a separate issue? Or will this issue be monitored here?

Furthermore, I would like to ask a question related to the use of the permission sharing. 
What is the best way to share a Realm with multiple users, but with the possiblility to revoke a shared Realm, while keeping the data available for all users to which it was already shared? 

How could the option to revoke be best implemented? And would it be better to let all users work in the shared Realm (which should then be synchronized with all user's (owned) Realms in case of a potential revoke)? Or would it be a better option to always keep the users owned Realms in sync with the shared Realm?

Any ideas how this should be implemented? Hopefully, some basic guidelines how to use the functionality will become available as well.

Many thanks.

**Comment by jpsim** [comment]
> A solution seemed to re-install the server everytime an empty database is required. Since I prefer to refresh the database several times during development this is not ideal. Should I submit this as a separate issue? Or will this issue be monitored here?

That shouldn't be required and I don't think we're aware of any such issues causing this, so yes, filing a new issue with exact steps to get ROS to this state would be much appreciated.

**Comment by jpsim** [comment]
> Furthermore, I would like to ask a question related to the use of the permission sharing.
> What is the best way to share a Realm with multiple users, but with the possiblility to revoke a shared Realm, while keeping the data available for all users to which it was already shared?

I think the best way to do this is to make a `SyncPermissionOffer` and resolve that for all users for which it should be applied by having each recipient create a `SyncPermissionOfferResponse`. The sender can then observe the permission Realm and the `SyncPermission` objects contained within it, then create a `SyncPermissionChange` with the user ID whose permissions should be changed (i.e. revoked).

**Comment by Taco55** [comment]
Thanks

I will file the first issue as a separate issue.

With respect to your answer. This sounds as a good way to handle the revoke (i.e. permission change indeed)l. Do you also have a suggestion how to keep all data in sync and maintain it after such a permission change? 
When a user creates a PermissionOfferResponse, the data in his/her owned Realm could be added to the shared Realm, or the data in the shared Realm could be distributed to the user-owned Realms. In the latter scenario, a permission change lets all users keep the data but the data is stored redundant in multiple Realms. 
Is there a "best" approach in this scenario?

**Comment by jpsim** [comment]
The "best" approach would be to compartmentalize your data, not copy it out into multiple Realms, since those copies are entirely separate and not linked in any way.

**Comment by Taco55** [comment]
That is true, but in case read/write access of the shared Realm is removed, all receiving users do not have access to "their" data anymore. Hence, the data should be copied to multiple Realms anyway, just before the permission change becomes effective.

However, before I ask too much questions on this, I will first implement it in the way you suggested and will see how I handle the data persistency after a potential permission change.

With respect to the main issue of this thread, since the bug that was causing this is solved successfully the issue can be closed. I filed the remaining issue as a new issue.

**Comment by jpsim** [comment]
Thanks for the thoughtful discussion, @Taco55!
